### PR TITLE
Finished resource relation implementation

### PR DESCRIPTION
**Pull request by pksunkara**
There are a few comments I want to make. So, please read them in the diff.

### CODE REVIEW

File: src/RelationParser.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+namespace snowcrash {
+
+    /** Link Relation matching regex */
+    const char* const RelationRegex = "^[[:blank:]]*[Rr]elation[[:blank:]]*:" SYMBOL_IDENTIFIER "?$";
+
+    /**
+     *  Relation Section Processor
+     */
+    template<>
+    struct SectionProcessor<Relation> : public SectionProcessorBase<Relation> {
```
Comment (pksunkara) [inline src/RelationParser.h]: I think we need to do something about this `SectionProcessor`. If I have

```
typedef std::string Relation;
typedef std::string Asset;
```

Then, defining `SectionProcessor<Asset>` and `SectionProcessor<Relation>` gives me redefinition errors. Which is why I had to do `struct Relation` earlier.

File: src/RelationParser.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#ifndef SNOWCRASH_RELATIONPARSER_H
+#define SNOWCRASH_RELATIONPARSER_H
+
+#include "SectionParser.h"
+#include "RegexMatch.h"
+
+namespace snowcrash {
+
+    /** Link Relation matching regex */
+    const char* const RelationRegex = "^[[:blank:]]*[Rr]elation[[:blank:]]*:" SYMBOL_IDENTIFIER "?$";
```
Comment (pksunkara) [inline src/RelationParser.h]: @zdne Are we going to be satisfied with `SYMBOL_IDENTIFIER` here? Or do we need to make a new regex `LINK_RELATION_IDENTIFIER`?

File: src/RelationParser.h (hunk 1 above)
Comment (klokane) [inline src/RelationParser.h]: It is due to both has same type. If you requre you have to make new type by inheritance

Comment (pksunkara) [inline src/RelationParser.h]: Yes, so, we need to refactor the `SectionProcessor` a bit later.

Comment (klokane) [inline src/RelationParser.h]: I start some refactoring there, 
https://github.com/apiaryio/snowcrash/compare/master...klokane:milestone-keywords

hope there will be time to finish :)  

At least we should introduce type traits to `SectionProcessor`

**Comment by klokane** [comment]
merged

File: src/RelationParser.h (hunk 2 above)
Comment (zdne) [inline src/RelationParser.h]: @pksunkara 

So according to [RFC 5988](https://tools.ietf.org/html/rfc5988) a relation type should be

```
  relation-types = relation-type
                 | <"> relation-type *( 1*SP relation-type ) <">
  relation-type  = reg-rel-type | ext-rel-type
  reg-rel-type   = LOALPHA *( LOALPHA | DIGIT | "." | "-" )
  ext-rel-type   = URI
```

So Snow Crash should really conform to this. Also in respect to what @klokane mentions it should be implemented with https://github.com/apiaryio/snowcrash/milestones/Keywords%20&%20Identifiers in mind. 

That is:

1. parser should warn when it encounters `+ Relation` anywhere except an action section
2. parser should recognize any `+ [Rr]elation\s+([.])* and then work with the capture group

Does this answer your question?

Comment (pksunkara) [inline src/RelationParser.h]: Got it. I will fix the regex in another PR tomorrow.

Comment (pksunkara) [inline src/RelationParser.h]: > Does this answer your question?

That wasn't my question at all. But @klokane understood it and addressed my concern.

**Pull request merged**